{"expireTime":9007200904342771000,"key":"transformer-remark-markdown-html-3cbbea9130985651fd87671609450238--","val":"<h1>Promise.any()</h1>\n<p><code>Promise.any()</code> takes an iterable of <a href=\"../promise\"><code>Promise</code></a> objects and, as soon as one of the promises in the iterable fulfills, returns a single promise that resolves with the value from that promise. If no promises in the iterable fulfill (if all of the given promises are rejected), then the returned promise is rejected with an <a href=\"../aggregateerror\"><code>AggregateError</code></a>, a new subclass of <a href=\"../error\"><code>Error</code></a> that groups together individual errors. Essentially, this method is the opposite of <a href=\"all\"><code>Promise.all()</code></a>.</p>\n<h2>Syntax</h2>\n<pre><code>Promise.any(iterable);\n</code></pre>\n<h3>Parameters</h3>\n<p><code>iterable</code>\nAn <a href=\"../../iteration_protocols#the_iterable_protocol\">iterable</a> object, such as an <a href=\"../array\"><code>Array</code></a>.</p>\n<h3>Return value</h3>\n<ul>\n<li>An <strong>already rejected</strong> <a href=\"../promise\"><code>Promise</code></a> if the iterable passed is empty.</li>\n<li>An <strong>asynchronously resolved</strong> <a href=\"../promise\"><code>Promise</code></a> if the iterable passed contains no promises.</li>\n<li>A <strong>pending</strong> <a href=\"../promise\"><code>Promise</code></a> in all other cases. This returned promise is then resolved/rejected <strong>asynchronously</strong> (as soon as the stack is empty) when any of the promises in the given iterable resolve, or if all the promises have rejected.</li>\n</ul>\n<h2>Description</h2>\n<p>This method is useful for returning the first promise that fulfills. It short-circuits after a promise fulfills, so it does not wait for the other promises to complete once it finds one. Unlike <a href=\"all\"><code>Promise.all()</code></a>, which returns an <em>array</em> of fulfillment values, we only get one fulfillment value (assuming at least one promise fulfills). This can be beneficial if we need only one promise to fulfill but we do not care which one does. Note another difference: This method rejects upon receiving an <em>empty iterable</em>, since, truthfully, the iterable contains no items that fulfill.</p>\n<p>Also, unlike <a href=\"race\"><code>Promise.race()</code></a>, which returns the first <em>settled</em> value (either fulfillment or rejection), this method returns the first <em>fulfilled</em> value. This method will ignore all rejected promises up until the first promise that fulfils.</p>\n<h3>Fulfillment</h3>\n<p>The returned promise is fulfilled with <strong>the first</strong> resolved value (or non-promise value) in the iterable passed as the argument, whether or not the other promises have rejected.</p>\n<ul>\n<li>If a nonempty <em>iterable</em> is passed, and <strong>any</strong> of the promises fulfill, or are not promises, then the promise returned by this method is fulfilled asynchronously.</li>\n</ul>\n<h3>Rejection</h3>\n<p>If all of the passed-in promises reject, <code>Promise.any</code> asynchronously rejects with an <a href=\"../aggregateerror\"><code>AggregateError</code></a> object, which extends <a href=\"../error\"><code>Error</code></a>, and contains an <code>errors</code> property with an array of rejection values.</p>\n<ul>\n<li>If an empty iterable is passed, then the promise returned by this method is rejected synchronously. The rejected reason is an <code>AggregateError</code> object whose <code>errors</code> property is an empty array.</li>\n</ul>\n<h2>Examples</h2>\n<h3>First to fulfil</h3>\n<p><code>Promise.any()</code> resolves with the first promise to fulfil, even if a promise rejects first. This is in contrast to <a href=\"race\"><code>Promise.race()</code></a>, which resolves or rejects with the first promise to settle.</p>\n<pre><code>const pErr = new Promise((resolve, reject) => {\n  reject(\"Always fails\");\n});\n\nconst pSlow = new Promise((resolve, reject) => {\n  setTimeout(resolve, 500, \"Done eventually\");\n});\n\nconst pFast = new Promise((resolve, reject) => {\n  setTimeout(resolve, 100, \"Done quick\");\n});\n\nPromise.any([pErr, pSlow, pFast]).then((value) => {\n  console.log(value);\n  // pFast fulfils first\n})\n// expected output: \"Done quick\"\n</code></pre>\n<h3>Rejections with AggregateError</h3>\n<p><code>Promise.any()</code> rejects with an <a href=\"../aggregateerror\"><code>AggregateError</code></a> if no promise fulfils.</p>\n<pre><code>const pErr = new Promise((resolve, reject) => {\n  reject('Always fails');\n});\n\nPromise.any([pErr]).catch((err) => {\n  console.log(err);\n})\n// expected output: \"AggregateError: No Promise in Promise.any was resolved\"\n</code></pre>\n<h3>Displaying the first image loaded</h3>\n<p>In this example, we have a function that fetches an image and returns a blob. We use <code>Promise.any()</code> to fetch a couple of images and display the first one available (i.e. whose promise has resolved).</p>\n<pre><code>function fetchAndDecode(url) {\n  return fetch(url).then(response => {\n    if(!response.ok) {\n      throw new Error(`HTTP error! status: ${response.status}`);\n    } else {\n      return response.blob();\n    }\n  })\n}\n\nlet coffee = fetchAndDecode('coffee.jpg');\nlet tea = fetchAndDecode('tea.jpg');\n\nPromise.any([coffee, tea]).then(value => {\n  let objectURL = URL.createObjectURL(value);\n  let image = document.createElement('img');\n  image.src = objectURL;\n  document.body.appendChild(image);\n})\n.catch(e => {\n  console.log(e.message);\n});\n</code></pre>"}